### ISSUE TITLE
RuntimeWarning: coroutine 'Redis.close' was never awaited

### ISSUE THREAD

**Issue by brianmaissy** [aio-libs/aioredis-py#1086]
I'm using aioredis in a pytest fixture, (using pytest-asyncio), and just upgraded to 2.0.0.

At the end of my test run, I'm getting the following warning:
```
/home/brian/repos/medigator/venv/lib/python3.8/site-packages/aioredis/client.py:1047: RuntimeWarning: coroutine 'Redis.close' was never awaited
  pass
Object allocated at (most recent call last):
  File "/home/brian/repos/medigator/venv/lib/python3.8/site-packages/aioredis/client.py", lineno 1045
    loop.run_until_complete(self.close())
```

I'm not sure exactly why it's happening, maybe it has something to do with the way pytest-asyncio is managing the loop, but it isn't really critical (I'm not saying aioredis is doing anything wrong).

I was wondering if it would be possible to add the `if conn` check that `close()` does to `__del__()` also, such that if I make sure to manually clean it up beforehand, the `close()` coroutine won't get created at `__del__` time.

Happy to open a PR if it would help.

**Comment by Olegt0rr** [comment]
@brianmaissy, since v2.0 you shouldn't close the Redis
https://aioredis.readthedocs.io/en/latest/migration/#cleaning-up

**Comment by abrookins** [comment]
I'm seeing this too and haven't fixed it yet. I'll take a closer look!

**Comment by krieghan** [comment]
I've been seeing this as well (on 2.0.0).  It seems to be related to the following code in client.py, specifically (I'm guessing) the create_task case:

    def __del__(self):
        try:
            loop = asyncio.get_event_loop()
            if loop.is_running():
                loop.create_task(self.close())
            else:
                loop.run_until_complete(self.close())
        except Exception:
            pass

I'm at a loss as to why it's not being awaited, though.  The errors I'm seeing at the end of my test suite look like this:

aioredis/client.py:1047: RuntimeWarning: coroutine 'Redis.close' was never awaited
sys:1: RuntimeWarning: coroutine 'Connection.disconnect' was never awaited

**Comment by Andrew-Chen-Wang** [comment]
Try `asyncio.run_coroutine_threadsafe` https://docs.python.org/3/library/asyncio-task.html#asyncio.run_coroutine_threadsafe Not sure if the coroutine is even put into queue for execution (you'll prob need to monitor the connections in redis-cli).

**Comment by krieghan** [comment]
> Try `asyncio.run_coroutine_threadsafe` https://docs.python.org/3/library/asyncio-task.html#asyncio.run_coroutine_threadsafe Not sure if the coroutine is even put into queue for execution (you'll prob need to monitor the connections in redis-cli).

Following up on this, I'm afraid using run_coroutine_threadsafe did not, in itself, fix or change the issue.  I'll continue looking at it to see if I can figure out specifically what's going on.

**Comment by symisc** [comment]
Any updates on this issue?

**Comment by jar3b** [comment]
> Any updates on this issue?

Not an update but I use non-optimal workaround with aiohttp in my code (`aioredis==2.0.0`):

```python
async def redis_init(app):
    app['rc'] = aioredis.from_url("redis://127.0.0.1")
    del type(app['rc']).__del__

async def redis_close(app):
    await app['rc'].close()
    
def run():
   app = web.Application()
   app.on_startup.append(redis_init)
   app.on_cleanup.append(redis_close)
```

This code cleanups Redis connection on exit but skips `__del__` call which produces this `never awaited` warining. The `__del__` method is useless if you using aiohttp because `.close()` was actually never called.

**Comment by Andrew-Chen-Wang** [comment]
Again, at least in django, we manually close any cache backend manuallly in a post-request hook. I'm +1 on just removing the `__del__` method entirely and have the user manually call a disconnect method.

**Comment by DemX86** [comment]
I have the same problem.
So far I am using the workaround suggested by jar3b but looking forward to clear solution.

**Comment by Andrew-Chen-Wang** [comment]
@DemX86 please check master as we've merged a potential solution.

**Comment by DemX86** [comment]
> @DemX86 please check master as we've merged a potential solution.
That's great, thanks a lot!
That master with the solution commit will be released soon with a new package version, right?

**Comment by seandstewart** [comment]
> Soon is iffy at the moment. @seandstewart thoughts on just releasing current version as patch? Incremental patching would help more people to slowly migrate.

Yeah I think that's a good idea. The migration to the latest redis client has numerous big changes so we'll want to have a separate release cycle and not block patches to the current version. We just need to make sure these changes make it into the new version as well.

**Comment by brianmaissy** [comment]
So this should be resolved by #1227?

**Comment by Andrew-Chen-Wang** [comment]
Yes. Closing for now. If it appears again, we'll reopen this issue.
